firestore/bulkwriter.go

```diff
@@ -0,0 +1,332 @@
+// Each call to Create, Update, Set, and Delete can return a value and error.
+type BulkWriter struct {
+	database        string             // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	start           time.Time          // when this BulkWriter was started; used to calculate qps and rate increases
+	vc              *vkit.Client       // internal client
+	isOpen          bool               // flag that the BulkWriter is closed
+	maxOpsPerSecond int                // number of requests that can be sent per second
+	docUpdatePaths  map[string]bool    // document paths with corresponding writes in the queue
+	cancel          context.CancelFunc // context to send cancel message
+	bundler         *bundler.Bundler   // handle bundling up writes to Firestore
+	retryBundler    *bundler.Bundler   // handle bundling up retries of writes to Firestore
+	ctx             context.Context    // context for canceling BulkWriter operations
+	rateLock        sync.Mutex         // guards increases to rate of open sends; used in increaseRate()
+}
+
+// newBulkWriter creates a new instance of the BulkWriter. This
+// version of BulkWriter is intended to be used within go routines by the
+// callers.
+func newBulkWriter(ctx context.Context, c *Client, database string) *BulkWriter {
+	ctx, cancel := context.WithCancel(withResourceHeader(ctx, c.path()))
+
+	var rateLock sync.Mutex
+
+	bw := &BulkWriter{
+		database:        database,
+		start:           time.Now(),
+		vc:              c.c,
+		isOpen:          true,
+		maxOpsPerSecond: defaultStartingMaximumOpsPerSecond,
+		docUpdatePaths:  make(map[string]bool),
+		cancel:          cancel,
+		bundler:         &bundler.Bundler{},
+		retryBundler:    &bundler.Bundler{},
+		ctx:             ctx,
+		rateLock:        rateLock,
+	}
+
+	bw.bundler = bundler.NewBundler(bulkWriterOperation{}, bw.send)
+	bw.bundler.HandlerLimit = bw.maxOpsPerSecond
+	bw.bundler.BundleCountThreshold = maxBatchSize
+	bw.bundler.BundleByteLimit = 0 // unlimited size
+
+	bw.retryBundler = bundler.NewBundler(bulkWriterOperation{}, bw.send)
+	bw.retryBundler.BundleCountThreshold = retryMaxBatchSize
+	bw.bundler.BundleByteLimit = 0 // unlimited size
+
+	return bw
+}
+
+// End sends all enqueued writes in parallel and closes the BulkWriter to new requests.
+// After calling End(), calling any additional method automatically returns
+// with an error. This method completes when there are no more pending writes
+// in the queue.
+func (b *BulkWriter) End() {
+	b.Flush()
+	b.cancel()
+	b.isOpen = false
+}
+
+// Flush commits all writes that have been enqueued up to this point in parallel.
+// This method blocks execution.
+func (bw *BulkWriter) Flush() {
+	// Ensure that the backlogQueue is empty
+	bw.bundler.Flush()
+	bw.retryBundler.Flush()
+}
+
+// Status gets the current open or closed state of the BulkWriter.
+func (bw *BulkWriter) Status() BulkWriterStatus {
```

Comment:
Is the idea that this could be more than just open and closed? If not this could just return a bool